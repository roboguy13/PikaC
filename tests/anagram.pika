%generate equal
-- %generate isAnagram

Sll : layout[-x](List);
Sll Nil := emp;
Sll (Cons head tail) :=
  exists +nxt.
  x :-> head **
  (x+1) :-> nxt **
  Sll tail [nxt];

equal : Sll -> Sll -> Bool;
equal Nil Nil := True;
equal (Cons a as) (Cons b bs) | (a == b) := (equal as bs);
equal (Cons a as) (Cons b bs) | not (a == b) := False;
equal Nil (Cons b bs) := False;
equal (Cons a as) Nil := False;

%test "equal1" Bool: equal (Cons 1 (Cons 2 (Nil [Sll]) [Sll]) [Sll]) (Cons 2 (Cons 3 (Nil [Sll]) [Sll]) [Sll]);
%test "equal2" Bool: equal (Cons 1 (Cons 2 (Nil [Sll]) [Sll]) [Sll]) (Nil [Sll]);
%test "equal3" Bool: equal (Cons 2 (Cons 3 (Nil [Sll]) [Sll]) [Sll]) (Cons 2 (Cons 3 (Nil [Sll]) [Sll]) [Sll]);

