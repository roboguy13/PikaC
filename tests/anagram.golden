*** Running test equal1
0

*** Running test equal2
0

*** Running test equal3
1

*** Running test isAnagram1
1

*** Running test isAnagram1
0

